lib/parser/context.rb

```diff
@@ -0,0 +1,47 @@
+module Parser
+  # Context of parsing that is represented by a stack of scopes.
+  class Context
+    attr_reader :stack
+
+    def initialize
+      @stack = []
+      freeze
+    end
+
+    def push(state)
+      @stack << state
+    end
+
+    def pop
+      @stack.pop
+    end
+
+    def reset
+      @stack.clear
+    end
+
+    def in_class?
+      @stack.last == :class
+    end
+
+    def in_sclass?
+      @stack.last == :sclass
+    end
+
+    def in_def?
```

Comment:
We already have a `Parser::Base.in_def?`; I think you should replace that with the new context machinery, and update all the older parsers as well.

Reply:
This flag is used to check that the parser is currently **somewhere** in `def` (to prevent class definition in the method body), not directly in `def`, so it's similar more to checking that the stack includes a `:def` symbol. Should I change the `Context` class to have methods like `in_def?` and `directly_in_def?` (and same stuff for other contexts)? Do you see any better names for these methods?

Reply:
Take a look at how `parse*.y` handle `@def_level`, they save and restore it to achieve the same behavior as here. Unless I'm missing something. Can you show the difference with some code example, please?